docs/proposals/memory-management.md

```diff
@@ -0,0 +1,185 @@
+
+2. Swapping is turned off by default. Kubelet does not track or manage swap files or partitions.
+
+3. A container in a pod with restart policy set to `Always` will be restarted continuously even if it OOMs. This is alleviated with the crash loop backoff feature. To improve node stability, that pod should be evicted and run on a different node.
+
+4. Other than the Guaranteed class, containers in other classes don't have a memory limit.
+A `Best-effort` container can use a lot of memory and result in high tail latencies for `Guaranteed` and `Burstable` class containers.
+Memory allocations will hang until the kernel can free up memory, by identifying and killing a process in a container in a lower class.
+
+5. Kubelet does not act on OOM kills.
+A container can choose to continue running when one or more processes in the container were killed by the kernel OOM killer.
+Because of this, the kubelet cannot deterministically associate container death with OOMs.
+This affects higher level systems like auto-scaler which requires reliable OOM signals to update memory limits.
+
+6. In the event of an OOM kill due to system memory pressure, the kubelet cannot identify the container of the killed process.
+
+7. All burstable containers that exceed their request are equally likely to be killed. Ideally, the job that exceeds the most should be killed.
+
+8. Memory usage of system daemons is unrestricted. This results in varying amount of memory availability on the node.
+
+
+##### How to address these limitations?
+
+We can improve memory management by making use of [kernel memcg features](https://www.kernel.org/doc/Documentation/cgroups/memory.txt), and by modifying the current system behavior. Some of these improvements are required in the short-term. The rest are either complex or require significant changes to the system.
+
+###### What can be done right away?
+
+1. Cleanup the entire container whenever a OOM kill is detected in a container.
+Users should be able to override the default behavior for containers that have an in-built process manager that can handle OOM kills.
+
+2. Set memory soft limits on Burstable containers to their request.
+This should result in "Burstable" containers being pushed to their requested memory limit under system memory pressure.
+
+3. Kubelet should expose `available` memory in addition to `capacity` and `usage`.
+The scheduler can then schedule containers based on `available` instead of `capacity`, which is the case as of now.
+To begin with `available` can be the difference between `capacity` and the total memory assigned to all system processes.
+Since we do not restrict the resource usage of system daemons, the `available` resources can differ over time.
+Another alternative is to create a shadow pod that will help statically reserve resources for system components.
```

Comment:
Maybe link to here: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/node.md#node-capacity